Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean

```diff
@@ -0,0 +1,89 @@
+
+noncomputable section
+suppress_compilation
+
+open FreeMonoid in
+instance instNoZeroDivisorsMonoidAlgebraFreeMonoid {R : Type*} [Semiring R] [NoZeroDivisors R]
+    {κ : Type*} [DecidableEq κ] : NoZeroDivisors (MonoidAlgebra R (FreeMonoid κ)) where
+  eq_zero_or_eq_zero_of_mul_eq_zero := fun hxy =>
+      or_iff_not_and_not.mpr <| not_and.mpr <| fun hx hy => not_not.mpr hxy <| by
+    haveI : DecidableEq (FreeMonoid κ) := inferInstanceAs (DecidableEq (List κ))
+    have max_length {x : MonoidAlgebra R (FreeMonoid κ)} (hx : x ≠ 0) :
+        ∃ w ∈ x.support, ∀ u ∈ x.support, u.length ≤ w.length :=
+      ⟨(x.support.toList.argmax (List.length ∘ FreeMonoid.toList)).get <|
+          Option.ne_none_iff_isSome.mp <| fun h => hx <| Finsupp.support_eq_empty.mp <|
+            Finset.toList_eq_nil.mp <| List.argmax_eq_none.mp <| h,
+        Finset.mem_toList.mp <| List.argmax_mem <| Option.get_mem _,
+        fun _ hu => List.le_of_mem_argmax (Finset.mem_toList.mpr hu) (Option.get_mem _)⟩
+    have ⟨xmax, hxmax_mem, hxmax_spec⟩ := max_length hx
+    have ⟨ymax, hymax_mem, hymax_spec⟩ := max_length hy
+    refine Finsupp.support_nonempty_iff.mp ⟨xmax * ymax, ?_⟩
+    rewrite [MonoidAlgebra.mul_def]
+    unfold Finsupp.sum
+    rewrite [← Finset.sum_product', Finsupp.mem_support_iff, ← Finsupp.applyAddHom_apply,
+      map_sum, Finset.sum_congr rfl (fun _ _ => by
+        rw [Finsupp.applyAddHom_apply, MonoidAlgebra.single_apply]),
+      Finset.sum_ite, Finset.sum_const_zero, add_zero,
+      Finset.sum_eq_single_of_mem (xmax, ymax) ?mem (fun p => ?zero)]
+    case mem
+    · rewrite [Finset.mem_filter, Finset.mem_product]
+      exact ⟨⟨hxmax_mem, hymax_mem⟩, rfl⟩
+    case zero
+    · rewrite [Finset.mem_filter, Finset.mem_product, ne_eq, Prod.ext_iff.not]
+      refine fun ⟨⟨hmem₁, hmem₂⟩, heq⟩ hne => False.elim <| hne <| List.append_inj heq ?_
+      apply And.left
+      rewrite [← toList.apply_eq_iff_eq, toList_mul, toList_mul] at heq
+      rewrite [← add_eq_add_iff_eq_and_eq (hxmax_spec _ hmem₁) (hymax_spec _ hmem₂),
+        ← List.length_append, ← List.length_append]
+      exact congrArg List.length heq
+    exact mul_ne_zero
+      (Finsupp.mem_support_iff.mp hxmax_mem) (Finsupp.mem_support_iff.mp hymax_mem)
+
+namespace TensorAlgebra
+
+instance instNoZeroDivisors {R : Type*} [CommSemiring R] [NoZeroDivisors R]
+    {M : Type*} [AddCommMonoid M] [Module R M] [Module.Free R M] :
+      NoZeroDivisors (TensorAlgebra R M) where
+  eq_zero_or_eq_zero_of_mul_eq_zero :=
+    fun {x y} hxy => or_iff_not_and_not.mpr <| not_and.mpr <| fun hx hy => not_not.mpr hxy <| by
+      have ⟨⟨κ, b⟩⟩ := ‹Module.Free R M›
+      haveI : DecidableEq κ := Classical.decEq κ
+      suffices FreeAlgebra.equivMonoidAlgebraFreeMonoid ((equivFreeAlgebra b) x) *
+          FreeAlgebra.equivMonoidAlgebraFreeMonoid ((equivFreeAlgebra b) y) ≠ 0 by
+        rwa [← map_mul, ← map_mul, AddEquivClass.map_ne_zero_iff,
+          AddEquivClass.map_ne_zero_iff] at this
```

Comment:
Is there a missing intermediate result about `FreeAlgebra` here?

Reply:
Maybe! Here I used the maximality of the lengths of `xmax ∈ x.support` and `ymax ∈ y.support` to show that `xmax * ymax ∈ (x * y).support`. I'm not sure how the proof would go for `MonoidAlgebra R G` without some such length/degree/grade/valuation on `G`.

Reply:
My claim here is that you are effectively using the fact that the `FreeAlgebra` has no zero divisors.

Reply:
(I assume you weren't referring only to the part you highlighted? That is just `(e x) * (e y) = 0 ↔ e (x * y) = 0` for an `e: AlgEquiv`, by an easy combination of existing lemmas, done twice.)

Reply:
> My claim here is that you are effectively using the fact that the `FreeAlgebra` has no zero divisors.

Sorry, my last message wasn't in reply to the above. Messages don't seem to be popping up in a timely manner.

But I think I coincidentally addressed your point? Those lines don't require `NoZeroDivisors`, if I'm not mistaken.

Reply:
See the commit I just pushed to see what I meant :)

Reply:
I updated the bulleted list in the commit comment.